sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,280 @@
+        neighbours_proba *= neighbours_mask
+
+        neighbours_term = np.einsum("ij,ijkl", neighbours_proba, outer)
+
+        assert neighbours_term.shape == (n_features, n_features)
+        grad -= neighbours_term
+
+        function_value = 0
+        if loss == 'l1':
+            function_value = p_i.sum()
+        elif loss == 'kl':
+            function_value = np.log(p_i).sum()
+
+        grad = 2 * np.dot(L, grad)
+
+        return [-function_value, -grad.flatten()]
+
+    return oracle
+
+
+def nca_semivectorized_oracle(X, y, n_components, loss, threshold=0.0):
+    n_samples, n_features = X.shape
+
+    def oracle(L):
+        L = L.reshape((n_components, n_features))
+
+        Lx = np.dot(X, L.T)  # n_samples x n_components
+        assert Lx.shape == (n_samples, n_components)
+
+        grad = np.zeros((n_features, n_features))
+        function_value = 0
+
+        for i in range(n_samples):
+            A = Lx[i, :] - Lx  # n_samples x n_components
+            assert A.shape == (n_samples, n_components)
+
+            logp = -(A * A).sum(axis=1)  # n_samples
+            logp[i] = -np.inf
+            logp -= sp.misc.logsumexp(logp)
+            assert logp.shape == (n_samples, )
+
+            p = np.exp(logp)  # n_samples
+            assert p.shape == (n_samples, )
+
+            class_neighbours = y == y[i]
+            p_i = (p * class_neighbours).sum()
+
+            if loss == 'l1':
+                samples_proba = p * p_i
+            elif loss == 'kl':
+                samples_proba = p
+            mask = samples_proba > threshold  # n_samples
+
+            Xij = X[i] - X[mask, :]  # n_relevant x n_features
+            Xij_outer = (Xij[:, :, np.newaxis] * Xij[:, np.newaxis, :])
+            grad += np.einsum("i,ijk", samples_proba[mask], Xij_outer)
+
+            if loss == 'l1':
+                neighbours_proba = p * (y == y[i])
+            elif loss == 'kl':
+                neighbours_proba = p / p_i * (y == y[i])
+            mask = neighbours_proba > threshold  # n_samples
+
+            Xij = X[i] - X[mask, :]  # n_relevant x n_features
+            Xij_outer = (Xij[:, :, np.newaxis] * Xij[:, np.newaxis, :])
+            grad -= np.einsum("i,ijk", neighbours_proba[mask], Xij_outer)
+
+            if loss == 'l1':
+                function_value += p_i
+            elif loss == 'kl':
+                function_value += np.log(p_i)
+
+        grad = 2 * np.dot(L, grad)
+
+        return [-function_value, -grad.flatten()]
+
+    return oracle
+
+
+def optimize_nca(X, y, learning_rate, n_components, loss, n_init, max_iter,
+                 method, tol, random_state, cache_size, verbose):
+    n_samples, n_features = X.shape
+
+    rng = np.random.RandomState(random_state)
+    best_value = np.inf
+    best_L = None
+
+
+    threshold = 1e-3 / np.abs(X).max() ** 2
+
+    outer_products_size = (n_samples * n_features) ** 2
+    memory_permits = outer_products_size < cache_size
+    if memory_permits:
```

Comment:
For consistency with the rest of scikit-learn, I'd call it `method`, not `scheme`.  I don't think `cache_size` (which should probably be `memory_size`) is a good idea:
  * It's a new way to parametrize things, that nothing else in scikit-learn uses.
  * Novice users won't know how to set it. 
  * It doesn't do what it says. It doesn't actually reserve a cache of that size. Advanced users who expect to be able to control memory usage will be misled. For example, if the data is large enough and `cache_size` is small enough, the code may well end up using more memory than `cache_size` is set to.

+1 for `method={'fully_vectorized' | 'semi_vectorized' | 'stochastic'}`. I see @eickenberg's point that all these methods have a right to exist, but I wouldn't overwhelm the user with options (and the developers with more code to maintain) unless `fully_vectorized` is clearly better than `semi_vectorized` in some scenarios.  Maybe if the data is small enough for the 4d tensor to fit in memory, then the loop won't be terribly slow either. (And it could run more times in parallel for the same memory cost.) It's just an idea, and it won't be relevant until we can fairly compare the three methods. 

And I'd suggest defaulting to `stochastic`, depending on how much memory and time `semi_vectorized` would take on, say, digits or faces. We try to keep estimators sensible when called with default arguments. People like to play around with stuff before reading instructions and understanding what stuff does. What do you think?